tq/transfer_queue.go

```diff
@@ -165,12 +196,266 @@ func (q *TransferQueue) Add(t Transferable) {
 		q.transferables[t.Oid()] = t
 		q.trMutex.Unlock()
 	} else {
-		tracerx.Printf("already transferring %q, skipping duplicate", t)
 		q.trMutex.Unlock()
+
+		tracerx.Printf("already transferring %q, skipping duplicate", t.Oid())
 		return
 	}
 
-	q.batcher.Add(t)
+	q.incoming <- t
+}
+
+// collectBatches collects batches in a loop, prioritizing failed items from the
+// previous before adding new items. The process works as follows:
+//
+//   1. Create a new batch, of size `q.batchSize`, and containing no items
+//   2. While the batch contains less items than `q.batchSize` AND the channel
+//      is open, read one item from the `q.incoming` channel.
+//      a. If the read was a channel close, go to step 4.
+//      b. If the read was a Transferable item, go to step 3.
+//   3. Append the item to the batch.
+//   4. Sort the batch by descending object size, make a batch API call, send
+//      the items to the `*adapterBase`.
+//   5. Process the worker results, incrementing and appending retries if
+//      possible.
+//   6. If the `q.incoming` channel is open, go to step 2.
+//   7. If the next batch is empty AND the `q.incoming` channel is closed,
+//      terminate immediately.
+//
+// collectBatches runs in its own goroutine.
+func (q *TransferQueue) collectBatches() {
+	var closing bool
+	batch := q.makeBatch()
+
+	for {
+		for !closing && !batch.IsFull(q.batchSize) {
+			t, ok := <-q.incoming
+			if !ok {
+				closing = true
+				break
+			}
+
+			batch = append(batch, t)
+		}
+
+		// Before enqueuing the next batch, sort it's items by size
+		// (largest first) so that workers can process the largest
+		// transfers first.
+		sort.Sort(sort.Reverse(batch))
```

Comment:
No worries, I wasn't quite clear with my reasoning here. The idea is to reduce the likelihood of a single worker getting tied up processing one very large item at the end of a batch while others are sitting idle. Since enqueuing the next batch is contingent on processing the current batch completely, idle workers should be treated as something we need to minimize as much as possible.

Specifically, the scenario I'm thinking about is a batch of 100 items where the first 99 are 5MB and the last item is 2GB. Processed in sequence, the workers will quickly finish transferring the small items. Towards the very end of the sequence, one worker will be sitting transferring a very large object, while the remaining workers are sitting doing nothing.

If the 2GB item is processed first, it ties a worker up for the duration of the transfer as it always would, but it allows the other workers to tear through the small items, minimizing the chance that a small number of large transfers will prolong the duration of the batch.

The sort makes this property much less noticeable, as does a larger number of workers, or a more evenly sized workload.

I went ahead and clarified this in b9aa0986.